File: 0380-Insert Delete GetRandom O(1)/0380-Insert Delete GetRandom O(1).cpp
class RandomizedSet {
public:
    /** Initialize your data structure here. */
    RandomizedSet() {
        
    }
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    bool insert(int val) {
        if (indexTable.find(val) != indexTable.end()) {
            return false;
        }
        
        indexTable[val] = nums.size();
        nums.push_back(val);
        return true;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    bool remove(int val) {
        if (indexTable.find(val) == indexTable.end()) {
            return false;
        }
        
        int index = indexTable[val];
        int last = nums.back();
        indexTable[last] = index;
        nums[index] = last;
        nums.pop_back();
        indexTable.erase(val);
        return true;
    }
    
    /** Get a random element from the set. */
    int getRandom() {
        return nums[rand() % nums.size()];
    }
    
private:
    vector<int> nums;
    unordered_map<int, int> indexTable;
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * bool param_1 = obj.insert(val);
 * bool param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */
 

File: 0301-0400/0380-Insert Delete GetRandom O(1)/0380-Insert Delete GetRandom O(1).py
import random
class RandomizedSet:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.nums = []
        self.indexTable = {}

    def insert(self, val: int) -> bool:
        """
        Inserts a value to the set. Returns true if the set did not already contain the specified element.
        """
        if val in self.indexTable:
            return False
        self.indexTable[val] = len(self.nums)
        self.nums.append(val)
        return True

    def remove(self, val: int) -> bool:
        """
        Removes a value from the set. Returns true if the set contained the specified element.
        """
        if val not in self.indexTable:
            return False
        index = self.indexTable[val]
        last = self.nums[-1]
        self.indexTable[last] = index
        self.nums[index] = last
        self.nums.pop()
        del self.indexTable[val]
        return True

    def getRandom(self) -> int:
        """
        Get a random element from the set.
        """
        return random.choice(self.nums)


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
